feat(e2e test): add basic e2e test

matwld opened the pull request:
Add a basic e2e test
- also change return value of deploy to 0 to match other commands

Run the test with `npm run e2e-test`

pr is in too parts because I accidentally pushed an altered head instead of making a a new commit

matwld committed:
feat(e2e test): add basic e2e test

Add a basic e2e test
- also change return value of deploy to 0 to match other commands

matwld committed:
feat(e2e test): add basic e2e test

Add a basic e2e test
- also change return value of deploy to 0 to match other commands

matwld committed:
chore(eslint): resolve eslint issues

Resolve eslint issues

johnnyw-aws commented on src/libs/tmdk/package.json:
```diff
@@ -9,6 +9,7 @@
   "scripts": {
     "test": "./node_modules/.bin/jest --silent",
     "integ-test": "./node_modules/.bin/jest --silent -c jest.config.integration.js",
+    "e2e-test": "./node_modules/.bin/jest -c jest.config.e2e.js",
```
instead of creating a new `e2e-test` can we just udpdate `integ-test`?

also since integ-test was made silent should e2e-test be made silent as well?

matwld replied:
We discussed with AppKit engineers that integ tests will be mocked through jest and verify api calls rather than validate state. I feel this test is better labeled as an e2e test. I agree having 3 types of tests is confusing, but I think the classifications are:
Unit tests: verify simple edge cases and method returns
Integ tests: verify api calls and expected behavior
E2E test: simulate user interaction and verify interaction with AWS

johnnyw-aws replied:
ok, I think we can delete `integ-test` for now if it means something specific in AppKit and call these "functional-tests" instead

matwld replied:
Got it, will update. So for now our current testing naming is:
Unit tests: verify api calls and edge cases
Functional tests: verify e2e user experience and AWS interaction

johnnyw-aws commented on src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts:
```diff
@@ -0,0 +1,492 @@
+
+  // 0. Get account info
+  console.log("Initializing AWS client.");
+  initDefaultAwsClients({ region: `${constants.region}` });
```
should region be typed to avoid this string interpolation?

matwld replied:
I think I just copied this from somewhere else, should have removed the interpolation and just pulled region directly. Will update

johnnyw-aws commented on src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts:
```diff
@@ -0,0 +1,492 @@
+  console.log("Initializing AWS client.");
+  initDefaultAwsClients({ region: `${constants.region}` });
+  const accountId = (await aws().getCurrentIdentity())["accountId"];
+  const twinMakerRoleName = constants.workspaceId + "-" + accountId + "-role";
+  const twinMakerPolicyName =
+    twinMakerRoleName + twinMakerPermissionPolicySuffix;
+  const twinMakerRoleArn =
+    "arn:aws:iam::" + accountId + ":role/" + twinMakerRoleName;
+  const workspaceS3BucketName = constants.workspaceId + "-" + accountId;
+  const workspaceS3BucketArn = "arn:aws:s3:::" + workspaceS3BucketName;
+  const scene1S3Location =
+    "s3://" + workspaceS3BucketName + "/" + constants.scene1FileName;
+  const model1S3Location =
+    "s3://" + workspaceS3BucketName + "/" + constants.model1FileName;
+  const model2S3Location =
+    "s3://" + workspaceS3BucketName + "/" + constants.model2FileName;
```
can we use string interpolation to simplify this?

matwld replied:
Yes, will update

johnnyw-aws commented on src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts:
```diff
@@ -0,0 +1,492 @@
+  const model2S3Location =
+    "s3://" + workspaceS3BucketName + "/" + constants.model2FileName;
+
+  // 1. Clean up pre-existing resources, if any
```
to avoid accidental deletion of a workspace that had issues and would like to debug, can we instead create a new  workspace (e.g. timestamp-suffixed) each time (since we should be fully cleaning up at end anyway)

we can also include something to help fully delete a workspace as you have here (or extend nuke to do so) if we end up with cruft workspaces to cleanup from integ test failures

matwld replied:
Sure, will add

johnnyw-aws commented on src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts:
```diff
@@ -0,0 +1,492 @@
+    );
+    await aws().s3.deleteBucket({ Bucket: workspaceS3BucketName });
+  } catch (e) {
+    console.log("S3 bucket does not exist, as expected.");
```
do we need to handle the case where the first file doesn't exist but the rest do?

matwld replied:
No, s3 delete object commands do not throw exception if the object does not exist

johnnyw-aws commented on src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts:
```diff
@@ -0,0 +1,492 @@
+      PolicyName: twinMakerPolicyName,
+      PolicyDocument: twinMakerPolicyString,
+    });
+    await delay(10000); // allow role to propagate
```
can we avoid blind waits? (or make TODO to revisit)

matwld replied:
Sure, will add TODO to verify role creation

johnnyw-aws commented on src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e-iam.ts:
```diff
@@ -0,0 +1,69 @@
+        "s3:GetBucketLocation",
+      ],
+      Resource: [
+        "s3ArnStar", // replaced with workspace bucket
```
maybe make these more distinctive as something to be replaced? (e.g. `___PLACEHOLDER_FOR_WORKSPACE_BUCKET___`)

matwld replied:
Will do

johnnyw-aws commented on src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e-constants.ts:
```diff
@@ -0,0 +1,49 @@
+export const localResourcesDir = path.join(__dirname, "basic-e2e-resources");
+export const workspaceId = "tmdk-e2e-test-workspace";
+export const region = "us-east-1";
+export const tmdkDirectory = "/tmp/e2e-tmdk-test";
```
just curious if this could make the integ test not work on certain platforms, e.g. Windows (can add to backlog if so)

matwld replied:
path.join should use the correct delimiter, but this is a good point for the tmp folder I use for tmdk. I'll update that to use path.join as well
https://nodejs.org/api/path.html

johnnyw-aws commented on src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts:
```diff
@@ -0,0 +1,492 @@
+  } as Arguments<init.Options>;
+  expect(await init.handler(argv2)).toBe(0);
+
+  // 5. Validate tmdk definition
```
would it be simpler to compare against a pre-saved tmdk definition?

matwld replied:
Yes, will add

johnnyw-aws commented on src/libs/tmdk/e2e-tests/basic-e2e/basic-e2e.test.ts:
```diff
@@ -0,0 +1,492 @@
+  expect(scene2ModelNames).toContain(
+    constants.model2FileName.replace(".glb", "")
+  );
+  // Verify Models
```
should we instead compare against the `basic-e2e-resources` ? (general comment for anything in S3 we should just be uploading)

matwld replied:
This is specifically checking whether the scene definition contains the correct model name. I could have used a const to store the model name but it was easier to just strip the file extension from the already saved file name

johnnyw-aws reviewed (commented):
thanks for adding this! overall looks good to me, just want to confirm some of the areas to see if ways to simplify

matwld commented:
Makes sense, I definitely lost steam on simplifying parts as I went on, will make changes and push a commit to resolve. Last concern is whether this should be labeled e2e test or integ test. I am fine with either, both make sense it seems only to matter to how granular we want to be with testing classification

matwld committed:
chore(e2e test): address pr comments

Address pr comments by:
- copy region directly when initializing aws client
- use string interpolation instead of concatenation
- add unix timestamp suffix to workspaceId
- add TODO to wait for role creation
- rename placeholder for s3 bucket in policy
- change paths used by test to be os agnostic
- use pre-saved tmdk definition to validate init success

matwld committed:
fix(functional test): rename e2e test to functional

- rename e2e test to functional
- remove integ tests
- use eslint fix to resolve eslint issues

matwld committed:
chore(functional test): silence test output

- silence test logging
- remove jest config for integration test
